### PR TITLE
#167392449 Implement Search Functionality

### PR DESCRIPTION
**Pull request by kechyy**
### What does this PR do?
- Implement search functionality

#### Description of Task to be completed?
- update user profile model with firstname and lastname
- create search functionality API controller
- create tag model
- create articletag model
- write unit tests to test the search functionality

#### How should this be manually tested?
- clone the repo
- run npm install
- start the server with npm run start:dev
- test on postman http://localhost:3000/api/v1/search?search='searchValue'

#### Any background context you want to provide?
- N/A

#### What are the relevant pivotal tracker stories?
- [#167392449](https://www.pivotaltracker.com/story/show/167392449)

#### Questions:
- N/A

### CODE REVIEW

**Review by timi-codes — CHANGES_REQUESTED** [review]
Good Job ladies 😂. I'm aware this PR is close to perfect 😇but I just have to request this changes because you forgot to document your API's.😎

**Comment by Onyimatics** [comment]
> Good Job ladies 😂. I'm aware this PR is close to perfect 😇but I just have to request this changes because you forgot to document your API's.😎

 😂Thank you @timi-codes . We've added the documentation. Kindly re-review and do the needful. 😂 😂

**Review by timi-codes — APPROVED** [review]

**Review by codexempire — APPROVED** [review]
Nice one 👍 great algorithm

File: src/test/search.spec.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    expect(res.body.data.length).to.equal(1);
+    expect(res.body.data[0].user_score).to.be.within(0, 1);
+    expect(res.body.data[0].title_score).to.be.within(0, 1);
+    expect(res.body.data[0].tag_score).to.be.within(0, 1);
+  });
+
+  it('should search for articles by tag', async () => {
+    const firstUser = await createUser();
+    const firstArticle = await createArticle(
+      await generateArticle({ authorId: firstUser.id })
```
Comment (segunolalive) [inline src/test/search.spec.js]:
```

const article = await generateArticle({ authorId: firstUser.id });
article.title = 'bush gehs';
const secondarticle = await generateArticle({ authorId: firstUser.id });
secondarticle.title = 'zzzz';

const firstArticle = await createArticle(article);

// send 'bush gehs' as your search query;

now you can assert that the title_scrore === 1


```

Comment (segunolalive) [reply src/test/search.spec.js]: create a mock that you know will not match the query in any way and assert that it didn't match. That's how you know your search doesn't just match everything in the db

**Review by segunolalive — CHANGES_REQUESTED** [review]
Good work. Attend to the comments.

Consider adding pagination to this.

File: src/test/search.spec.js (hunk 1 above)
Comment (Onyimatics) [reply src/test/search.spec.js]: All right, we will do that.

**Comment by Onyimatics** [comment]
> Good work. Attend to the comments.
> 
> Consider adding pagination to this.

Okay, we will attend to them. Thank you.

File: src/test/search.spec.js (hunk 1 above)
Comment (kechyy) [reply src/test/search.spec.js]: > create a mock that you know will not match the query in any way and assert that it didn't match. That's how you know your search doesn't just match everything in the db

Okay, we'll implement the changes as requested. 

Thank you

**Review by tolumide-ng — APPROVED** [review]
Great job...welldone guys 👍

**Pull request merged**